lib/update-fetch.js

```diff
@@ -0,0 +1,205 @@
+  a Promise.
+*/
+function fetchBuildList() {
+  return new Promise(function(resolve, reject) {
+    // Fetch the list of available builds
+    request.get(buildsListURL, function(err, response, body) {
+
+      // If there wasn't an issue with the request
+      if (requestRespHelper(resolve, reject, err, response)) {
+        // Resolve with the parsed data
+        try {
+          resolve(JSON.parse(body));
+        }
+        // If the parse failed, reject
+        catch (err) {
+          reject(err);
+        }
+      }
+    });
+  });
+}
+
+/* 
+  Accepts a build name and attempts to fetch
+  the build images from the server. Returns build contents
+  in a Promise
+*/
+function fetchBuild(version) {
+  return new Promise(function(resolve, reject) {
+    logs.info('Beginning update download. This could take a couple minutes..');
+    return fetchSHAForBuildVersion(version)
+      .then(function startDownload(sha) {
+
+        var build = {
+          firmware: undefined,
+          openwrt: undefined,
+        };
+
+        // Fetch the list of available builds
+        extract.on('entry', function(header, stream, callback) {
+          // The buffer to save incoming data to
+          // The filename of this entry
+          var filename = path.basename(header.name);
+          // This entry is the openwrt binary
+          if (filename === openWRTBuild) {
+            // Save incoming data to the openwrt buffer 
+            streamToBuffer(stream, function(err, buffer) {
+              build.openwrt = buffer;
+              callback();
+            });
+          }
+          // This entry is the firmware binary
+          else if (filename === firmwareBuild) {
+            // Save incoming data to the firmware buffer
+            streamToBuffer(stream, function(err, buffer) {
+              build.firmware = buffer;
+              callback();
+            });
+          } else {
+            callback();
+          }
+        });
+
+        extract.once('finish', function() {
+          if (!build.firmware.length || !build.openwrt.length) {
+            return reject(new Error('Fetched binary wasn\'t formatted properly.'));
+          } else {
+            logs.info('Download complete!');
+            return resolve(build);
+          }
+        });
+        // Make a request to our build server for this particular sha
+        var req = request.get(urljoin(firmwareBuildURL, sha + '.tar.gz'));
+        // unzip and extract the binary tarball
+        req.pipe(gunzip).pipe(extract);
+      });
+  });
+}
+
+function requestRespHelper(resolve, reject, err, response) {
+  // If there was an error making the request, reject
+  if (err) {
+    reject(err);
+    return false;
+  }
+
+  // If there was an issue with the server endpoint, reject
+  else if (response.statusCode !== 200) {
+    reject(new Error('Invalid status code on build server request: ' + response.statusCode));
+    return false;
+  }
+  // Otherwise
+  else {
+    return true;
```

Comment:
This function is bad news :\ It's used as a [predicate](https://github.com/tessel/t2-cli/pull/231/files#diff-1ff7e3d75491fb02338063a4a45559fdR24), which is expected to return a boolean value (or some kind of coercible value), but may handle the rejection? 

Suggest: 

```js
function reviewResponse(response) {
  var outcome = {
    success: true, 
  };

  // If there was an issue with the server endpoint, reject
  if (response.statusCode !== 200) {
    outcome.success = false;
    outcome.reason = 'Invalid status code on build server request: ' + response.statusCode;
  }

  return outcome;
}
```

Then above: 

```js
function fetchBuildList() {
  return new Promise(function(resolve, reject) {
    // Fetch the list of available builds
    request.get(buildsListURL, function(err, response, body) {
      if (err) {
        return reject(err);
      }

      var outcome = reviewResponse(response);

      // If there wasn't an issue with the request
      if (outcome.success) {
        // Resolve with the parsed data
        try {
          resolve(JSON.parse(body));
        }
        // If the parse failed, reject
        catch (err) {
          reject(err);
        }
      } else {
        reject(outcome.reason);
      }
    });
  });
}
```

I'm sure that can even be further refined